File: Source/python/filesorter.py
import os
import shutil
from collections import Counter

def organize_files_by_extension(folder_path):
    if folder_path == None:
        return
    files = [f for f in os.listdir(folder_path) if os.path.isfile(os.path.join(folder_path, f))]
    extensions = []

    for file in files:
        file_path = os.path.join(folder_path, file)
        name, ext = os.path.splitext(file)
        ext = ext.lower()

        if not ext:
            ext = '[no extension]'
            target_folder = os.path.join(folder_path, 'NO_EXTENSION')
        else:
            ext_name = ext.lstrip('.').upper() + 's'  # e.g., '.png' -> 'PNGs'
            target_folder = os.path.join(folder_path, ext_name)

        # Track extensions for counting
        extensions.append(ext if ext else '[no extension]')

        # Make sure the target folder exists
        os.makedirs(target_folder, exist_ok=True)

        # Move the file
        shutil.move(file_path, os.path.join(target_folder, file))

    # Count and print results
    extension_counts = Counter(extensions)
    print("Extension counts:")
    for ext, count in extension_counts.items():
        print(f"{ext}: {count}")

File: source/filesorter.js
import {
    promises as fs
} from "fs";
import path from "path";

/**
 * Organize files in a folder by their extension.
 * Creates subfolders (e.g., "PNGs", "TXTs") or "NO_EXTENSION".
 */
export async function organizeFilesByExtension(folderPath) {
    if (!folderPath) {
        console.log("Error: File path null");
        return;
    }

    // Get only files (skip subdirectories)
    const entries = await fs.readdir(folderPath, {
        withFileTypes: true
    });
    const files = entries.filter((e) => e.isFile()).map((e) => e.name);

    const extensions = [];

    for (const file of files) {
        const filePath = path.join(folderPath, file);
        let ext = path.extname(file).toLowerCase();
        let targetFolder;

        if (!ext) {
            ext = "[no extension]";
            targetFolder = path.join(folderPath, "NO_EXTENSION");
        } else {
            const extName = ext.slice(1).toUpperCase() + "s"; // e.g., ".png" → "PNGs"
            targetFolder = path.join(folderPath, extName);
        }

        extensions.push(ext || "[no extension]");

        await fs.mkdir(targetFolder, {
            recursive: true
        });
        await fs.rename(filePath, path.join(targetFolder, file));
    }

    // Count and print results
    const counts = extensions.reduce((acc, ext) => {
        acc[ext] = (acc[ext] || 0) + 1;
        return acc;
    }, {});

    console.log("Extension counts:");
    for (const [ext, count] of Object.entries(counts)) {
        console.log(`${ext}: ${count}`);
    }
}

// Example usage
// (Uncomment the next line and set the folder you want)
// organizeFilesByExtension("C:/path/to/your/folder");